controller/api/destination/watcher/cluster_store.go

```diff
@@ -0,0 +1,270 @@
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	consts "github.com/linkerd/linkerd2/pkg/k8s"
+)
+
+type (
+	// ClusterStore indexes clusters in which remote service discovery may be
+	// performed. For each store item, an EndpointsWatcher is created to read
+	// state directly from the respective cluster's API Server. In addition,
+	// each store item has some associated and immutable configuration that is
+	// required for service discovery.
+	ClusterStore struct {
+		// Protects against illegal accesses
+		sync.RWMutex
+
+		k8sAPI               *k8s.API
+		store                map[string]remoteCluster
+		enableEndpointSlices bool
+		log                  *logging.Entry
+
+		// Function used to parse a kubeconfig from a byte buffer. Based on the
+		// kubeconfig, it creates API Server clients
+		decodeFn configDecoder
+	}
+
+	// remoteCluster is a helper struct that represents a store item
+	remoteCluster struct {
+		watcher *EndpointsWatcher
+		config  clusterConfig
+
+		// Used to signal shutdown to the associated watcher's informers
+		stopCh chan<- struct{}
+	}
+
+	// clusterConfig holds immutable configuration for a given cluster
+	clusterConfig struct {
+		TrustDomain   string
+		ClusterDomain string
+	}
+
+	// configDecoder is the type of a function that given a byte buffer, returns
+	// a pair of API Server clients. The cache uses this function to dynamically
+	// create clients after discovering a Secret.
+	configDecoder = func(data []byte, enableEndpointSlices bool) (*k8s.API, *k8s.MetadataAPI, error)
+)
+
+const (
+	clusterNameLabel        = "multicluster.linkerd.io/cluster-name"
+	trustDomainAnnotation   = "multicluster.linkerd.io/trust-domain"
+	clusterDomainAnnotation = "multicluster.linkerd.io/cluster-domain"
+)
+
+// NewClusterStore creates a new (empty) ClusterStore. It
+// requires a Kubernetes API Server client instantiated for the local cluster.
+//
+// When created, a pair of event handlers are registered for the local cluster's
+// Secret informer. The event handlers are responsible for driving the discovery
+// of remote clusters and their configuration
+func NewClusterStore(k8sAPI *k8s.API, enableEndpointSlices bool) (*ClusterStore, error) {
+	return newClusterStoreWithDecoder(k8sAPI, enableEndpointSlices, decodeK8sConfigFromSecret)
+}
+
+// newClusterStoreWithDecoder is a helper function that allows the creation of a
+// store with an arbitrary `configDecoder` function.
+func newClusterStoreWithDecoder(k8sAPI *k8s.API, enableEndpointSlices bool, decodeFn configDecoder) (*ClusterStore, error) {
+	cs := &ClusterStore{
+		store: make(map[string]remoteCluster),
+		log: logging.WithFields(logging.Fields{
+			"component": "cluster-store",
+		}),
+		enableEndpointSlices: enableEndpointSlices,
+		k8sAPI:               k8sAPI,
+		decodeFn:             decodeFn,
+	}
+
+	_, err := cs.k8sAPI.Secret().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Why is that we're not interested in Update events? Can you add a comment about that?